src/org/etools/j1939_84/controllers/part12/Part12Step12Controller.java

```diff
@@ -0,0 +1,211 @@
+                                                             })
+                                                             .collect(Collectors.toList());
+                      if (packets.isEmpty()) {
+                          addFailure("6.12.12.3.a - NACK received for DM7 PG from OBD ECU "
+                                  + module.getModuleName() + " for SP " + spn);
+                      }
+
+                  });
+
+        });
+
+        getDataRepository().getObdModuleAddresses().forEach(moduleAddress -> {
+            getDm58AndVerifyData(moduleAddress);
+        });
+
+    }
+
+    private void getDm58AndVerifyData(int moduleAddress) {
+
+        // 6.12.11.4 Actions3:
+        // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
+        // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
+        // preferred when available).
+        var nonRatFaultSps = getDataRepository().getObdModules()
+                                                .stream()
+                                                .flatMap(m -> m.getSupportedSPNs().stream())
+                                                .filter(supported -> !supported.supportsRationalityFaultData())
+                                                .collect(Collectors.toList());
+        if (nonRatFaultSps.isEmpty()) {
+            getListener().onResult("6.12.12.4.a - No SPs found that do NOT indicate support for DM58 in the DM24 response from "
+                    + Lookup.getAddressName(moduleAddress));
+        } else {
+            int requestSpn = nonRatFaultSps.stream()
+                                           .filter(SupportedSPN::supportsScaledTestResults)
+                                           .findFirst()
+                                           .orElseGet(() -> nonRatFaultSps.get(0))
+                                           .getSpn();
+
+            var packet = getCommunicationsModule().requestDM58(getListener(), moduleAddress, requestSpn)
+                                                  .requestResult()
+                                                  .getAcks()
+                                                  .stream()
+                                                  .findFirst()
+                                                  .orElse(null);
+
+            // 6.12.11.5 Fail/Warn criteria3:
+            // a. Fail if a NACK is not received
+            if (packet == null || packet.getResponse() != AcknowledgmentPacket.Response.NACK) {
+                addFailure("6.12.12.5.a - NACK not received for DM7 PG from OBD ECU "
+                        + Lookup.getAddressName(moduleAddress) + " for spn " + requestSpn);
+            }
+        }
+    }
+
+    private boolean areUnusedBytesPaddedWithFFh(DM58RationalityFaultSpData packet) {
+        Slot slot = J1939DaRepository.findSlot(packet.getSpn().getId(), packet.getSpnId());
```

Comment:
This is passing in the SPN Id twice.  I’m guessing one of these should be the a Slot number.